### PR TITLE
Feature implementation: environments

### PR DESCRIPTION
**Pull request by isinyaaa**
This aims at implementing the environments feature (roughly described in #114).

There are some non-related commits in this branch as I made small improvements to other parts of the code I had been investigating, and those should be separated in other PR's as needed.

Tasklist:
- [x] Add feature base, having a clearer idea of what its API should be doing
    
    For this I've mainly been thinking in the following behavior:
    ```bash
    kw env (-s | --switch) <name> [(-d | \--description) <description>] [--keep-local] [(-g | --config) <config-name>] - Switch to another (new) environment
    kw env (-l | --list) - List existing kw environments
    kw env (-c | --clean) {name|(-a | --all)} - Clean an environment
    kw env (-D | --destroy) {name|(-a | --all)} [-f | --force] - Destroy an environment
    ```
    
    - One of the main problems I have with this API is the `--keep-local` flag, for which I couldn't think of an easy implementation.
    
    - Also, another thing that should be discussed is some sort of integration between the config manager's metadata and this one's as they should be very much related (one compilation output per config/arch).
- [ ] Implement feature details and integrations
- [ ] Clean up details, review git history, and separate (likely) other PR's

### CODE REVIEW

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @isinyaaa ,

First of all, thanks a lot for working on this feature; this is a very welcome feature (I know @melissawen really wants something like this).

I started to review your code, but I stopped because we first need a plan. This is my suggestion:

## Commit 1: Prepare kw build to work with different env

We probably need to rework kw build code to make it more flexible in accepting the `O=PATH` option. We can fix it by removing some assumptions and setting some variables.

## Commit 2: Prepare kw deploy to work with different env

We probably need to make kw deploy more flexible to read different folders during the deploy. In particular, pay attention to modules_uinstall and run_kernel_install.

## Commit 3: Preparation for kw env --create

When creating an env, we first need to create the metadata related to the new env. Maybe we can create a new folder under the `.kw/` folder, something like this:

`.kw/env/<NAME>/`

Inside this folder, we can have:

* A kworkflow.config file associated with the current env.
* env.data: Any info related with the target env (maybe we can add the description here)

We can also have a new and straightforward file inside `.kw/` named `current_env.config` that looks like this:

```
NAME_X
```

Finally, we need to "teach" kw to read this file to know which env is currently in use.

## Commit 4: Implement kw --create

Now that we have the basic setup, let's implement the `--create` option. I would start with this behavior:

* By default, when we create a new env, we automatically switch to it.
* Add name and description

With this behavior in mind, I can see two major scenarios:

* New kernel repository (easy)

Let's say that the user just cloned a kernel repository, and everything is clean. In this case `O=output/dir` will work out of the box.

* Kernel repository already in use

Users already have a kernel repository that they regularly use with many object files and a config file (let's call it a polluted repository). In this case, we will need to alert users that we will migrate their current workflow to the env case, and they will probably need to rebuild their kernel.

In this case, we need to run some sort of migration from local files to start using env (unpolluted?). I suppose we can reuse the current kworkflow.config and config, but we will probably need to remove all the object files.

In summary, when users invoke `kw env --create ...` we are probably going to have these actions:

1. Create a new env folder and files inside .kw/
2. Migrate user to the new approach if necessary

## Commit 5: Implement kw --switch

This option should switch between different events. It should be straightforward at this point, but let's ensure:

1. Only switch to a valid env
2. Update the current env

## Commit 6: List env

## Commit 7: kw b --clean

We will need to implement a new wrapper inside kw build to clean things. This option should recognize the env.

## Commit 8: kw env --destroy

File: documentation/man/features/environment.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  provide any command option, kw will assume the list option.
+
+-c, \--clean <name>:
+  Get a config file based on the *<name>* and paste it in the current
+  directory. It pop-up a warning message because this operation override the
+  current **.config** file. The user can suppress this warning by using ``-f``
+  flag.
+
+-D, \--destroy <name>:
+  Remove config labeled with *<name>*. It pop-up a warning message because it
```
[the remark this answers is not in the document]
Comment (isinyaaa) [reply documentation/man/features/environment.rst:41]: Ah yes, the text wasn't updated yet! I'm sorry about that.

File: documentation/man/features/environment.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+will list all existing environments under the current project.
+
+OPTIONS
+=======
+-s, \--switch <name> [(-d | --description) <description>]:
+  The save option searches the current directory for a **.config** file to be
+  kept under the management of **kw**. The save option expects a name to identify
+  this version of the file. Additionally, users can add a description by
+  using ``-d`` flag. Finally, if the user tries to add the same name twice,
+  **kw** will issue a warning; ``-f`` will suppress this message.
```
[the remark this answers is not in the document]
Comment (isinyaaa) [reply documentation/man/features/environment.rst:28]: I always thinking about that just as I read your comments, it does make more sense!

**Comment by rodrigosiqueira** [comment]
Hi @isinyaaa , since you are not working in this PR directly, I'm going to close it.
